smithy-cli/src/main/java/software/amazon/smithy/cli/commands/ClasspathAction.java

```diff
@@ -133,13 +136,32 @@ private List<Path> resolveDependencies(
         long lastModified = smithyBuildConfig.getLastModifiedInMillis();
         DependencyResolver delegate = new FilterCliVersionResolver(SmithyCli.getVersion(), baseResolver);
         DependencyResolver resolver = new FileCacheResolver(getCacheFile(buildOptions, smithyBuildConfig),
-                                                            lastModified,
-                                                            delegate);
-        addConfiguredMavenRepos(smithyBuildConfig, resolver);
-        maven.getDependencies().forEach(resolver::addDependency);
+                lastModified,
+                delegate);
+
+        Set<MavenRepository> repositories = DependencyUtils.getConfiguredMavenRepos(smithyBuildConfig);
+        Set<String> dependencies = maven.getDependencies();
+
+        // Use the lock file dependencies and repositories if a lockfile exists
+        Optional<LockFile> lockFileOptional = DependencyUtils.loadLockfile();
+        if (lockFileOptional.isPresent()) {
+            if (lockFileOptional.get().getConfigHash() != DependencyUtils.configHash(dependencies, repositories)) {
+                throw new CliError(
+                        "Lockfile does not match configured dependencies. Re-lock dependencies or revert changes.");
+            }
+            LOGGER.fine(() -> "Lockfile found. Using pinned dependencies: "
+                    + lockFileOptional.get().getDependencyCoordinateSet());
+            dependencies = lockFileOptional.get().getDependencyCoordinateSet();
```

Comment:
Don't you need to overwrite repositories too?

Reply:
I dont think it is necessary. The repo URLs are checked via the config_hash so users cant actually change those without invalidating the lockfile. They could change the login info, but I think that is ok.

The dependency coordinates on the other hand could be dependency ranges so we need to replace those with the locked version for resolution.